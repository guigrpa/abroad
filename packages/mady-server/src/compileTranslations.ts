// @flow

import MessageFormat from 'messageformat';
import Terser from 'terser';
import type { Key, Translation } from './types';

// ==============================================
// Main
// ==============================================
const compileTranslations = async ({
  lang,
  keys,
  translations,
  fAlwaysIncludeKeysWithBraces = true,
  fMinify = false,
}: {
  lang: string;
  keys: Record<string, Key>;
  translations: Translation[];
  fAlwaysIncludeKeysWithBraces?: boolean;
  fMinify?: boolean;
}) => {
  // =====================================
  // Prepare messageFormatTranslations and markdownTranslations
  // =====================================
  const messageFormatTranslations: Record<string, string> = {};
  const markdownTranslations: Record<string, string> = {};

  // We must always include those keys that are not markdown and
  // include curly braces, even if there is no translation.
  // These initial translations may be modified later on
  if (fAlwaysIncludeKeysWithBraces) {
    Object.keys(keys).forEach((keyId) => {
      const key = keys[keyId];
      if (!key || key.unusedSince) return;
      if (!key.isMarkdown && key.text.indexOf('{') >= 0) {
        messageFormatTranslations[keyId] = keys[keyId].text;
      }
    });
  }

  // Copy non-deleted translations to either messageFormatTranslations
  // or markdownTranslations
  translations.forEach((translation) => {
    const { keyId } = translation;
    const key = keys[keyId];
    if (!key || key.unusedSince) return;
    if (key.isMarkdown) {
      markdownTranslations[keyId] = translation.translation;
    } else {
      messageFormatTranslations[keyId] = translation.translation;
    }
  });

  // =====================================
  // Compile translation function
  // =====================================
  const mf = new MessageFormat(lang);
  let fnTranslate = mf
    .compile(messageFormatTranslations)
    .toString('module.exports');
  fnTranslate += ';\n';
  Object.keys(markdownTranslations).forEach((keyId) => {
    fnTranslate += `module.exports['${keyId}'] = function() {
      return ${JSON.stringify(markdownTranslations[keyId])}
    };\n`;
  });

  // =====================================
  // Minify result
  // =====================================
  if (fMinify) fnTranslate = (await Terser.minify(fnTranslate)).code as string;

  return fnTranslate;
};

// ==============================================
// Public
// ==============================================
export default compileTranslations;
