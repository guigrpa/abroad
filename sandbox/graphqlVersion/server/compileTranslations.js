// @flow

import MessageFormat from 'messageformat';
import Terser from 'terser';
import { chalk } from 'storyboard';
import type {
  MapOf,
  StoryT,
  InternalKeyT,
  InternalTranslationT,
} from '../common/types';

export default function compileTranslations({
  lang,
  keys,
  translations,
  fAlwaysIncludeKeysWithBraces = true,
  fMinify = false,
  story,
}: {|
  lang: string,
  keys: MapOf<InternalKeyT>,
  translations: Array<InternalTranslationT>,
  fAlwaysIncludeKeysWithBraces: boolean,
  fMinify?: boolean,
  story: StoryT,
|}): string {
  const logPrefix = `Lang ${chalk.magenta.bold(lang)}`;

  story.info('compiler', `${logPrefix} Preparing translations...`);

  // =====================================
  // Prepare messageFormatTranslations and markdownTranslations
  // =====================================
  const messageFormatTranslations = {};
  const markdownTranslations = {};

  // We must always include those keys that are not markdown and
  // include curly braces, even if there is no translation
  if (fAlwaysIncludeKeysWithBraces) {
    Object.keys(keys).forEach(keyId => {
      const key = keys[keyId];
      if (!key.isMarkdown && key.text.indexOf('{') >= 0) {
        messageFormatTranslations[keyId] = keys[keyId].text;
      }
    });
  }

  // Copy non-deleted translations to either messageFormatTranslations
  // or markdownTranslations
  translations.forEach(translation => {
    const { keyId } = translation;
    const key = keys[keyId];
    if (!key || key.unusedSince) return;
    if (key.isMarkdown) {
      markdownTranslations[keyId] = translation.translation;
    } else {
      messageFormatTranslations[keyId] = translation.translation;
    }
  });
  story.trace('compiler', `${logPrefix} Translations prepared`, {
    attach: messageFormatTranslations,
  });

  // =====================================
  // Compile translation function
  // =====================================
  story.info('compiler', `${logPrefix} Precompiling...`);
  const mf = new MessageFormat(lang).setIntlSupport(true);
  const fnMessageFormat = mf.compile(messageFormatTranslations).toString();
  let fnTranslate = `
function anonymous() {
${fnMessageFormat}
}
module.exports = anonymous();
`;
  Object.keys(markdownTranslations).forEach(keyId => {
    fnTranslate += `module.exports['${keyId}'] = function() {
      return ${JSON.stringify(markdownTranslations[keyId])}
    };\n`;
  });
  story.trace('compiler', `${logPrefix} Precompiled`, { attach: fnTranslate });

  // =====================================
  // Minify result
  // =====================================
  if (fMinify) {
    story.info('compiler', `${logPrefix} Minifying...`);
    fnTranslate = Terser.minify(fnTranslate).code;
    story.trace('compiler', `${logPrefix} Minified`, { attach: fnTranslate });
  } else {
    story.info('compiler', `${logPrefix} Skipped minification`);
  }

  return fnTranslate;
}
