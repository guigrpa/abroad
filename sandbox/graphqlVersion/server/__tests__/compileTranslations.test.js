/* eslint-env jest */
import { merge } from 'timm';
import { mainStory } from 'storyboard';
import compileTranslations from '../compileTranslations';

const KEYS_WITHOUT_BRACES = {
  keyId1: {
    id: 'keyId1',
    context: 'context',
    text: 'Hello there!',
    sources: ['FILENAME'],
    unusedSince: null,
  },
  keyId2: {
    id: 'keyId2',
    context: 'context',
    text: 'Bye-bye!',
    sources: ['FILENAME'],
    unusedSince: null,
  },
  keyId5: {
    id: 'keyId5',
    context: 'context',
    text: 'Hello again!',
    sources: [],
    unusedSince: '2010-01-01T10:00:00Z',
  },
};

const KEYS_INCLUDING_SOME_WITH_BRACES = merge(KEYS_WITHOUT_BRACES, {
  keyId3: {
    id: 'keyId3',
    context: 'context',
    text: 'Hello {NAME}!',
    sources: ['FILENAME'],
    unusedSince: null,
  },
  keyId4: {
    id: 'keyId4',
    context: 'context',
    text: 'Bye-bye, {NAME}!',
    sources: ['FILENAME'],
    unusedSince: null,
  },
});

const TRANSLATIONS_ARRAY = [
  { keyId: 'keyId1', lang: 'es', translation: '¡Hola!' },
  { keyId: 'keyId4', lang: 'es', translation: '¡Adiós, {NAME}!' },
  { keyId: 'keyId5', lang: 'es', translation: '¡Hola de nuevo!' },
];

describe('compileTranslations', () => {
  it('should include translations for keys still being used', () => {
    const result = compileTranslations({
      lang: 'es',
      keys: KEYS_WITHOUT_BRACES,
      translations: TRANSLATIONS_ARRAY,
      story: mainStory,
    });
    expect(result).toMatchSnapshot();
  });

  it('should include keys with braces, even if no translation is available', () => {
    const result = compileTranslations({
      lang: 'es',
      keys: KEYS_INCLUDING_SOME_WITH_BRACES,
      translations: TRANSLATIONS_ARRAY,
      story: mainStory,
    });
    expect(result).toMatchSnapshot();
  });

  it('should compile a key with plural selection', () => {
    const result = compileTranslations({
      lang: 'es',
      keys: {
        ID: {
          id: 'ID',
          text: '{NUM, plural, one{1 hamburger} other{# hamburgers}}',
        },
      },
      translations: [
        {
          keyId: 'ID',
          lang: 'es',
          translation:
            '{NUM, plural, one{1 hamburguesa} other{# hamburguesas}}',
        },
      ],
      story: mainStory,
    });
    expect(result).toMatchSnapshot();
  });

  it('should compile a key with Markdown', () => {
    const result = compileTranslations({
      lang: 'es',
      keys: {
        ID: {
          id: 'ID',
          text: '# Title',
          isMarkdown: true,
        },
      },
      translations: [
        {
          keyId: 'ID',
          lang: 'es',
          translation: '# Título',
        },
      ],
      story: mainStory,
    });
    expect(result).toMatchSnapshot();
  });

  it('should allow minification', () => {
    const result = compileTranslations({
      lang: 'es',
      keys: KEYS_INCLUDING_SOME_WITH_BRACES,
      translations: TRANSLATIONS_ARRAY,
      fMinify: true,
      story: mainStory,
    });
    expect(result).toMatchSnapshot();
  });
});
