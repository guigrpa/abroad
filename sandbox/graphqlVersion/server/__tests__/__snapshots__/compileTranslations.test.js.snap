// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compileTranslations should allow minification 1`] = `"function anonymous(){return{keyId3:function(n){return\\"Hello \\"+n.NAME+\\"!\\"},keyId4:function(n){return\\"¡Adiós, \\"+n.NAME+\\"!\\"},keyId1:function(n){return\\"¡Hola!\\"}}}module.exports=anonymous();"`;

exports[`compileTranslations should compile a key with Markdown 1`] = `
"
function anonymous() {

return {
}
}
module.exports = anonymous();
module.exports['ID'] = function() {
      return \\"# Título\\"
    };
"
`;

exports[`compileTranslations should compile a key with plural selection 1`] = `
"
function anonymous() {
var es = function(n, ord) {
  if (ord) return 'other';
  return (n == 1) ? 'one' : 'other';
};
var number = function (value, name, offset) {
  if (!offset) return value;
  if (isNaN(value)) throw new Error('Can\\\\'t apply offset:' + offset + ' to argument \`' + name +
                                    '\` with non-numerical value ' + JSON.stringify(value) + '.');
  return value - offset;
};
var plural = function(value, offset, lcfunc, data, isOrdinal) {
  if ({}.hasOwnProperty.call(data, value)) return data[value];
  if (offset) value -= offset;
  var key = lcfunc(value, isOrdinal);
  if (key in data) return data[key];
  return data.other;
};

return {
  ID: function(d) { return plural(d.NUM, 0, es, { one: \\"1 hamburguesa\\", other: number(d.NUM, \\"NUM\\") + \\" hamburguesas\\" }); }
}
}
module.exports = anonymous();
"
`;

exports[`compileTranslations should include keys with braces, even if no translation is available 1`] = `
"
function anonymous() {

return {
  keyId3: function(d) { return \\"Hello \\" + d.NAME + \\"!\\"; },
  keyId4: function(d) { return \\"¡Adiós, \\" + d.NAME + \\"!\\"; },
  keyId1: function(d) { return \\"¡Hola!\\"; }
}
}
module.exports = anonymous();
"
`;

exports[`compileTranslations should include translations for keys still being used 1`] = `
"
function anonymous() {

return {
  keyId1: function(d) { return \\"¡Hola!\\"; }
}
}
module.exports = anonymous();
"
`;
